tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphFunction.java

```diff
@@ -0,0 +1,301 @@
+      outputNames.add(output.getKey());
+      outputList.add(output.getValue());
+    }
+
+    try (PointerScope scope = new PointerScope()) {
+      TF_Status status = TF_Status.newStatus();
+
+      //TODO can't get output names working, likely due to passing it wrong
+
+//      PointerPointer<CharPointer> outputNamesPointer = new PointerPointer<>(outputNames.size());
+//      outputNamesPointer.putString(outputNames.toArray(new String[0]));
+
+      try (Reference ref = body.ref()) {
+        TF_Function handle = TF_GraphToFunction(
+            ref.nativeHandle(),
+            name,
+            (byte) (appendHash ? 1 : 0),
+            -1,
+            null,
+            inputList.size(),
+            resolveToOutput(inputList),
+            outputList.size(),
+            resolveToOutput(outputList),
+            (byte[]) null,
+            null,
+            description,
+            status
+        );
+
+        status.throwExceptionIfNotOK();
+
+        return new NamedGraphFunction(handle.withDeallocator(), inputNames, outputNames);
+      }
+    }
+  }
+
+  /**
+   * Create a {@link NamedGraphFunction} from a {@link ConcreteFunction}.
+   *
+   * @param function the {@code ConcreteFunction} to create the function from
+   * @param appendHash whether to append a hash of the signature to the name
+   */
+  public static NamedGraphFunction create(ConcreteFunction function, boolean appendHash) {
+    Graph graph = function.graph();
+
+    Map<String, Operand<?>> inputs = new LinkedHashMap<>();
+    for (Entry<String, TensorInfo> input : function.signature().asSignatureDef().getInputsMap().entrySet()) {
+      inputs.put(input.getKey(), graph.output(input.getValue().getName()));
+    }
+
+    Map<String, Operand<?>> outputs = new LinkedHashMap<>();
+    for (Entry<String, TensorInfo> output : function.signature().asSignatureDef().getOutputsMap().entrySet()) {
+      outputs.put(output.getKey(), graph.output(output.getValue().getName()));
+    }
+
+    String description = function.signature().methodName();
+    if (description == null) {
+      description = function.signature().key();
+    }
+
+    return create(graph, function.signature().key(), appendHash, inputs, outputs, description);
+  }
+
+  TF_Function nativeHandle() {
+    if (nativeHandle.isNull()) {
+      throw new IllegalStateException("Function has been closed");
+    }
+    return nativeHandle;
+  }
+
+  /**
+   * Get the name of the function.
+   */
+  public String getName() {
+    try (PointerScope scope = new PointerScope()) {
+      return TF_FunctionName(nativeHandle()).getString();
+    }
+  }
+
+  /**
+   * Close the function
+   */
+  @Override
+  public void close() {
+    scope.close();
+  }
+
+  /**
+   * Get the native handle of the function's gradient, so that it can be attached to a Graph.  Not implemented yet.
+   *
+   * TODO implement
+   */
+  TF_Function gradNativeHandle() {
+    return null;
```

Comment:
Any idea how to implement this?

Reply:
I need to do some tests, but for Graph I'm thinking I'd call the existing gradient op on it, and then construct a function using the op list constructor.  Eager is up in the air until we get gradient tape support.